### PR TITLE
Gluon PReLU, ELU, SELU, Swish

### PR DESCRIPTION
**Pull request by szha**
## Description ##
Picking up #8912 (@joeddav), #9111 (@anjishnu)

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fix `LeakyReLU(act_type='prelu')`
- [x] `nn.PReLU` and test
- [x] `nn.ELU`, `nn.SELU`, `nn.Swish` and tests

## Comments ##
- Fixed operator `LeakyReLU(act_type='prelu')`, since it didn't have the proper parameter or logic.

### CODE REVIEW

File: python/mxnet/gluon/nn/activations.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        Initializer for the `embeddings` matrix.
+
+
+    Inputs:
+        - **data**: input tensor with arbitrary shape.
+
+    Outputs:
+        - **out**: output tensor with the same shape as `data`.
+    """
+    def __init__(self, alpha_initializer='zeros', *args):
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/activations.py]: Is 0 initialization standard?

File: python/mxnet/gluon/nn/activations.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+    Inputs:
+        - **data**: input tensor with arbitrary shape.
+
+    Outputs:
+        - **out**: output tensor with the same shape as `data`.
+    """
+    def __init__(self, **kwargs):
+        super(SELU, self).__init__(**kwargs)
+        self.scale = 1.0507009873554804934193349852946
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/activations.py]: self._scale

File: python/mxnet/gluon/nn/activations.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    Inputs:
+        - **data**: input tensor with arbitrary shape.
+
+    Outputs:
+        - **out**: output tensor with the same shape as `data`.
+    """
+
+    def __init__(self, beta=1.0, **kwargs):
+        super(Swish, self).__init__(**kwargs)
+        self.beta = beta
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/activations.py]: _beta

File: src/operator/leaky_relu-inl.h [hunk 4]
```diff
@@ -177,9 +182,20 @@ class LeakyReLUOp : public Operator {
       case leakyrelu::kPReLU: {
         weight = in_data[leakyrelu::kGamma].get<xpu, 1, real_t>(s);
         grad_weight = in_grad[leakyrelu::kGamma].get<xpu, 1, real_t>(s);
-        grad_weight = sumall_except_dim<1>(F<prelu_grad>(data) * grad);
-        gdata = F<mshadow_op::xelu_grad>(data, mshadow::expr::broadcast<1>(weight, data.shape_))
-                * grad;
+        if (weight.shape_[0] == 1) {
```
Comment (piiswrong) [inline src/operator/leaky_relu-inl.h]: Could you explain? I thought prelu was already supported

Comment (bradcar) [reply src/operator/leaky_relu-inl.h]: Since uyou are adding those I would suggest also adding a learnable ISRLU:
https://arxiv.org/abs/1710.09967

Comment (szha) [reply src/operator/leaky_relu-inl.h]: Existing PReLU had two problems:
1. gamma parameter was never documented and can't be passed in using kwargs.
2. it doesn't support scalar broadcast as was attempted in #8912

Comment (piiswrong) [reply src/operator/leaky_relu-inl.h]: If you meant to test for scalar than you should use `weight.shape_.Size()`

File: python/mxnet/gluon/nn/activations.py (hunk 1 above)
Comment (szha) [reply python/mxnet/gluon/nn/activations.py]: [Tensorflow/Keras uses zeros](https://www.tensorflow.org/api_docs/python/tf/keras/layers/PReLU#__init__), [Pytorch uses 0.25](http://pytorch.org/docs/master/nn.html?highlight=prelu#torch.nn.PReLU)

Do we have a constant initializer to achieve the latter?

File: src/operator/leaky_relu-inl.h (hunk 4 above)
Comment (szha) [reply src/operator/leaky_relu-inl.h]: It's equivalent because weight is `Tensor<xpu, 1>`. I think `Size()` is safer choice in case `weight` changes definition.

Comment (piiswrong) [reply src/operator/leaky_relu-inl.h]: Infershape sets it to data.shape[1].
When would weight's shape be (1,)?

Comment (szha) [reply src/operator/leaky_relu-inl.h]: If the weight parameter is shared across all axis, then the only one scalar value is shared everywhere, in which case the weight should be (1,)

Comment (piiswrong) [reply src/operator/leaky_relu-inl.h]: But infershape doesn't allow this

Comment (szha) [reply src/operator/leaky_relu-inl.h]: @bradcar sorry that I missed your comment earlier, and thanks for sharing your work. In this PR I'd like to first focus on wrapping up the previous two PRs for activations. Since you wrote the paper, would you like to implement that in mxnet?

Comment (szha) [reply src/operator/leaky_relu-inl.h]: There are two options, either writing it in Gluon by defining hybrid_forward in python, or extending the leaky relu operator in C for better performance.

**Comment by szha** [comment]
@piiswrong addressed the infer shape issue. Let me know if you have more comments.

File: python/mxnet/gluon/nn/activations.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    def __init__(self, alpha_initializer=initializer.Constant(0.25), *args):
+        super(PReLU, self).__init__(*args)
+        with self.name_scope():
+            self.alpha = self.params.get('alpha', shape=(1,), init=alpha_initializer)
+
+    def hybrid_forward(self, F, x, alpha):
+        return F.LeakyReLU(x, gamma=alpha, act_type='prelu', name='fwd')
+
+    def __repr__(self):
+        s = '{name}'
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/activations.py]: Can we handle these trivial cases in base class?

File: tests/python/unittest/test_gluon.py [hunk 6]
```diff
@@ -719,6 +719,42 @@ def test_inline():
...
+            scale, alpha = 1.0507009873554804934193349852946, 1.6732632423543772848170429916717
+            return scale * x if x >= 0 else alpha * mx.nd.exp(x) - alpha
+        return [selu(x_i) for x_i in x]
+
+    for test_point, ref_point in zip(selu(point_to_validate), selu(point_to_validate)):
+        assert test_point == ref_point
+
+    prelu = mx.gluon.nn.PReLU()
+    prelu.initialize()
+    x = point_to_validate.reshape((1, 1, 2))
```
Comment (piiswrong) [inline tests/python/unittest/test_gluon.py]: use a different input shape that can catch the infershape problem

File: src/operator/leaky_relu-inl.h [hunk 7]
```diff
@@ -225,7 +242,11 @@ class LeakyReLUProp : public OperatorProperty {
     const TShape &dshape = in_shape->at(leakyrelu::kData);
     if (dshape.ndim() == 0) return false;
     if (param_.act_type == leakyrelu::kPReLU) {
-      in_shape->at(leakyrelu::kGamma) = TShape(Shape1(dshape[1]));
+      const TShape &gshape = in_shape->at(leakyrelu::kGamma);
+      if (gshape.Size() != 1)
```
Comment (piiswrong) [inline src/operator/leaky_relu-inl.h]: gshape could be empty, in which case Size is undefined.
Also if gshape.Size is 1, it could be (1,1), which is invalid

File: python/mxnet/gluon/nn/activations.py (hunk 5 above)
Comment (szha) [reply python/mxnet/gluon/nn/activations.py]: Yes

File: src/operator/leaky_relu-inl.h (hunk 7 above)
Comment (szha) [reply src/operator/leaky_relu-inl.h]: So, I should check for both ndim and shape_[0] then. How do I check whether it’s undefined?

Comment (piiswrong) [reply src/operator/leaky_relu-inl.h]: if gshape is empty gshape.ndim() would be 0

**Comment by szha** [comment]
@piiswrong, I addressed latest comments. Let me know if any further change is needed.

**Comment by szha** [comment]
ping @piiswrong

**Comment by lvyong1943** [comment]
hello ,when i run the mlp.cpp,there is an err like this:
Symbol.ComposeKeyword argument name gamma not found.
Candidate arguments:
	[0]data
when it run  outputs[i] = LeakyReLU(string("act") + istr, fc, null_sym, LeakyReLUActType::kLeaky);
so,how can i solve this problem？
And I change the  Symbol null_sym to  Symbol null_sym("null_sym") to fix the err, this->blob_ptr_._Ptr is nullptr。

**Comment by szha** [comment]
@lvyong1943 https://github.com/apache/incubator-mxnet/pull/9662/files#diff-be63e28c2263dcd0e3a7116463607037R72

**Comment by lvyong1943** [comment]
@szha  I just run the mlp.cpp file, when it run this line：
 Executor* exe = new Executor(sym_out, ctx_dev, in_args, arg_grad_store,grad_req_type, aux_states); 
there is an err appeared in  incubator-mxnet-master\src\executor\graph_executor.cc    at    line 273 .  
      xs.emplace_back(NodeEntry{args[i], 0, 0});   
 The parameter ‘i’ out of range "args".

**Comment by bradcar** [comment]
@piiswrong @szha what is the status of 9662 and prelu working? When I naively put PReLU (into a hybrid block and look at this code it seems that PReLU only has one learnable alpha per layer.

**Comment by bradcar** [comment]
@piiswrong @szha what is the status of 9662 and prelu working? When I naively put PReLU (into a hybrid block (mxnet 1.2.0) and look at the source (activations.py) it seems that PReLU only has one learnable alpha per layer. Shouldn't each 'neuron' have its own learnable alpha?

**Comment by szha** [comment]
@bradcar the leaky relu operator in 'prelu' mode supports any broadcast-able alpha shapes. Since it's impossible to infer the shape of parameter until it sees the first input, we chose to put the simplest case in the constructor.

For your use case when you need more than one alpha parameters, you can simply use the operator.
